File: fl_backend/app/api/einladungen/schemas.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# The application's slice declares the delivery state, and one shape is stored at every home the
# register names (`app/api/zustellung/services.py :: ZIEL_PFADE`).
from app.api.bewerbungen.schemas import FLBewerbungZustellung, FLKontaktRolle
from app.shared.schemas.custom import CustomDateString, CustomNonEmptyString, CustomObjectId, CustomOptionalDateString
from app.shared.schemas.responses import BaseAPIResponse


class FLEinladungVersand(BaseModel):
    """The carrier a message about this invite is recorded in.

    A carrier rather than the record itself: `zustellung_pfad` composes a dotted path out of it, and
    the accepted send skips a row holding no carrier at all.
    """

    # Absent until something is mailed, and null never: the send writes the whole record
    # (`app/api/zustellung/services.py :: compose_ziel_zustellung_update`).
    zustellung: FLBewerbungZustellung | None = None


class FLEinladung(BaseModel):
    """One minted link as it is served -- and NO `token_hash`.

    A model declaring the hash would put every live link of the season on an admin read
    (`docs/backend/spec.md :: I275`).
    """

    id: CustomObjectId = Field(validation_alias="_id", serialization_alias="id")
    saison_id: CustomNonEmptyString
    team_id: CustomObjectId
    erstellt_am: CustomDateString
    # The bound actor, on no payload: a link and its `aktionen` row cannot then name two people.
    erstellt_von: CustomNonEmptyString
    # Required as a KEY and null while the invite stands: `uniq_einladung_live` reaches the rows
    # holding null, so a row that could omit it would fall outside the one-live-invite rule.
    widerrufen_am: CustomOptionalDateString
    versand: FLEinladungVersand | None = None


class FLEinladungMintResponse(BaseAPIResponse):
    """A freshly minted registration link, and the only answer that ever carries its raw value.

    Shown once for copying and mailed on a separate press; no later read recovers it from the hash
    (`docs/backend/spec.md :: I275`).
    """

    saison_id: CustomNonEmptyString
    team_id: CustomObjectId
    einladung_id: CustomObjectId
    token: CustomNonEmptyString
    erstellt_am: CustomDateString
    erstellt_von: CustomNonEmptyString


class FLEinladungWriteResponse(BaseAPIResponse):
    """The revoked row. Revoking keeps it, so a delivery event about its link still lands somewhere."""

    saison_id: CustomNonEmptyString
    team_id: CustomObjectId
    einladung_id: CustomObjectId


class FLEinladungResponse(BaseAPIResponse):
    """The team's live link for one season, or nothing, beside the window deciding what it opens.

    Declared here rather than in `app/api/teams/schemas.py`, whose endpoint answers it: moving it
    there is an import cycle through `app/api/bewerbungen/schemas.py`.
    """

    saison_id: CustomNonEmptyString
    team_id: CustomObjectId
    # Null where the team holds none: a link is minted on a press, so holding none is the state
    # every team starts in.
    einladung: FLEinladung | None
    # Composed here rather than derived by the reader, as `FLBewerbungFensterResponse` composes its
    # own: a link is live and opens nothing the moment the registration window shuts.
    laeuft: bool


class FLEinladungEmpfaenger(BaseModel):
    """One mailbox a link goes to, and the seat it is addressed as."""

    rolle: FLKontaktRolle
    vorname: CustomNonEmptyString
    email: CustomNonEmptyString


# Four ordinary states and a failure: `erzeugung_fehlgeschlagen` alone says the league failed that
# team, and a surface words it apart from the rest (`docs/backend/spec.md :: I282`).
FLEinladungVersandGrund = Literal[
    "austritt_eingetragen",
    "erzeugung_fehlgeschlagen",
    "kein_kontaktblock",
    "keine_bestaetigte_kontaktperson",
    "bereits_gesendet",
]


class _EinladungVersandZeile(BaseModel):
    """What the preview and the press both say about one team."""

    team_id: CustomObjectId
    # The junction row's own copy of the club's name, so the list reads as the season's teams do
    # rather than needing a second read of `teams`.
    team_name: CustomNonEmptyString
    empfaenger: list[FLEinladungEmpfaenger]
    uebersprungen: FLEinladungVersandGrund | None
    # Served rather than derived: a row carries no trace of the invitation the team holds, so a
    # caller cannot tell a team whose link this press kills from one that never had a link.
    ersetzt_link: bool


class FLEinladungVersandVorschauZeile(_EinladungVersandZeile):
    """One team as the preview reports it. It mints nothing, so it answers no link."""


class FLEinladungVersandZeile(_EinladungVersandZeile):
    """One team as the press left it."""

    # Null exactly where the team was skipped. The raw value is answered here and never again, a
    # link being recoverable from no read.
    einladung_id: CustomObjectId | None
    token: str | None


class FLEinladungVersandPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # False by default, so pressing twice mails nobody twice: a team whose live link already carries
    # a delivery record is skipped unless this says otherwise.
    erneut: bool = False


class FLEinladungVersandVorschauResponse(BaseAPIResponse):
    """Who the press would mail, over the same rule the press then performs."""

    saison_id: CustomNonEmptyString
    zeilen: list[FLEinladungVersandVorschauZeile]


class FLEinladungVersandResponse(BaseAPIResponse):
    """What the press did, per admitted team of the season."""

    saison_id: CustomNonEmptyString
    zeilen: list[FLEinladungVersandZeile]

File: fl_frontend/src/features/einladungen/schemas.ts
import z from "zod";

import { BaseAPIResponseSchema } from "@/core/schemas";
// The applications slice declares the width every minted token is bounded by, one mint function
// serving both flows (`fl_backend/app/api/bewerbungen/services.py :: mint_token`); a literal beside
// it is compared by nothing.
import { BEWERBUNG_TOKEN_MAX_LENGTH } from "@/features/bewerbungen/constants";
import { FLBewerbungZustellungSchema, FLKontaktRolleSchema } from "@/features/bewerbungen/schemas";
import { SAISON_ID_LENGTH } from "@/features/saisons/constants";
import { CustomDateStringSchema, CustomObjectIdStringSchema } from "@/shared/schemas";

/**
 * Which team's invite, in a season. Both ids travel in the PATH on the wire; they are carried here
 * because the action has to know whose invite it is minting.
 */
export const FLEinladungKeyPayloadSchema = z.object({
  team_id: CustomObjectIdStringSchema,
  saison_id: z.string().length(SAISON_ID_LENGTH, { error: "Bitte wähle eine Saison." }),
});
export type FLEinladungKeyPayload = z.infer<typeof FLEinladungKeyPayloadSchema>;

/**
 * **The raw link value is the panel's to hand back**: the store keeps a hash alone, so the mint's
 * own answer is the only place it exists, and a message composed after a reload would carry none.
 */
export const FLEinladungMailPayloadSchema = FLEinladungKeyPayloadSchema.extend({
  einladung_id: CustomObjectIdStringSchema,
  token: z
    .string()
    .trim()
    // German though the panel offers no box for it: a payload refused here reaches the log rather
    // than a field, and `fl_frontend/src/core/schemaGerman.test.ts` holds every payload to German.
    .nonempty({ error: "Diese Einladung nennt keinen Link." })
    .max(BEWERBUNG_TOKEN_MAX_LENGTH, { error: "Dieser Link ist zu lang." }),
});
export type FLEinladungMailPayload = z.infer<typeof FLEinladungMailPayloadSchema>;

/**
 * Mirrors `FLEinladungVersand`. **A carrier holding no `zustellung` is an invite NOBODY MAILED**,
 * which is a different fact from a message that went out and did not arrive.
 */
export const FLEinladungVersandSchema = z.object({
  // Nullable and never optional: the mint writes the carrier EMPTY, but FastAPI serialises the
  // default, so the key rides on every invitation with `null` in it. The three sibling mirrors of
  // this same backend model spell it the same way.
  zustellung: FLBewerbungZustellungSchema.nullable(),
});
export type FLEinladungVersand = z.infer<typeof FLEinladungVersandSchema>;

/**
 * Mirrors `FLEinladung` — the live invite as an administrator is shown it. **No `token_hash`**: the
 * credential is written as a hash the store never yields, so no read model carries one and neither
 * may this (`docs/backend/spec.md :: I144`'s twin).
 */
export const FLEinladungSchema = z.object({
  // The row a delivery report is filed against (`ziel_id`), which is why a read model carries it at
  // all: the mail press is a second press, and by then the mint's own answer is gone.
  id: CustomObjectIdStringSchema,
  saison_id: z.string(),
  team_id: CustomObjectIdStringSchema,
  erstellt_am: CustomDateStringSchema,
  // The administrator's own address, as `bewerbungen.entscheidung.von` is read from the bound actor.
  erstellt_von: z.string(),
  // Null on every row this read serves, the read finding the live one; the field is mirrored because
  // the stored row carries it and a reader of the model would otherwise think a revoke leaves none.
  widerrufen_am: CustomDateStringSchema.nullable(),
  versand: FLEinladungVersandSchema.nullable(),
});
export type FLEinladung = z.infer<typeof FLEinladungSchema>;

/** `null` where the team holds no live invite for the season, which is every team before the first mint. */
export const FLEinladungResponseSchema = BaseAPIResponseSchema.extend({
  saison_id: z.string(),
  team_id: CustomObjectIdStringSchema,
  einladung: FLEinladungSchema.nullable(),
  // The registration window's verdict, composed server-side as `FLBewerbungFensterResponse.laeuft`
  // is: a link expires with the window rather than on a date of its own.
  laeuft: z.boolean(),
});
export type FLEinladungResponse = z.infer<typeof FLEinladungResponseSchema>;

/**
 * What a mint answers, and the only place the raw link value exists outside the recipient's inbox:
 * the panel shows it once, for copying, and a reload cannot get it back.
 */
export const FLEinladungMintResponseSchema = BaseAPIResponseSchema.extend({
  saison_id: z.string(),
  team_id: CustomObjectIdStringSchema,
  einladung_id: CustomObjectIdStringSchema,
  token: z.string(),
  erstellt_am: CustomDateStringSchema,
  erstellt_von: z.string(),
});
export type FLEinladungMintResponse = z.infer<typeof FLEinladungMintResponseSchema>;

/** What the revoke answers: the row it closed. Nothing reverses it — the next link is a fresh mint. */
export const FLEinladungWriteResponseSchema = BaseAPIResponseSchema.extend({
  saison_id: z.string(),
  team_id: CustomObjectIdStringSchema,
  einladung_id: CustomObjectIdStringSchema,
});
export type FLEinladungWriteResponse = z.infer<typeof FLEinladungWriteResponseSchema>;

/**
 * **One payload for the preview AND the press**: a shape carrying the re-send choice to one of them
 * alone is how a page comes to list one set of teams and write to another.
 */
export const FLEinladungVersandPayloadSchema = z.object({
  // The season travels in the PATH on the wire; it is carried here because both calls have to name
  // the season they were pressed on.
  id: z.string().length(SAISON_ID_LENGTH, { error: "Bitte wähle eine Saison." }),
  // Optional because the endpoint defaults it, which on a REQUEST is what optional means: a body
  // without the key mails nobody twice, the failure a bulk control has and a per-team one does not.
  erneut: z.boolean().optional(),
});
export type FLEinladungVersandPayload = z.infer<typeof FLEinladungVersandPayloadSchema>;

/**
 * Mirrors `FLEinladungVersandGrund`. **Judged by the endpoint and never re-derived here**: a second
 * derivation is how a preview starts telling the administrator one thing while the press does another.
 */
export const FLEinladungVersandGrundSchema = z.enum(
  // The withdrawal FIRST, in the endpoint's own order: a skip naming the contact block instead
  // would send somebody to enter contacts for a team with nothing left to register for.
  [
    "austritt_eingetragen",
    "erzeugung_fehlgeschlagen",
    "erzeugung_ungewiss",
    "kein_kontaktblock",
    "keine_bestaetigte_kontaktperson",
    "bereits_gesendet",
  ],
  { error: "Diesen Grund gibt es nicht." },
);
export type FLEinladungVersandGrund = z.infer<typeof FLEinladungVersandGrundSchema>;

/**
 * Mirrors `FLEinladungEmpfaenger` — one MAILBOX the send would write to. A person holding two seats
 * is one row, named for the first seat they hold, because they get one message rather than two.
 */
export const FLEinladungEmpfaengerSchema = z.object({
  rolle: FLKontaktRolleSchema,
  vorname: z.string(),
  // Shapeless, as every stored mailbox is on a read: a value refused here would fail the whole list
  // rather than the one seat holding it (`docs/backend/spec.md :: I36`).
  email: z.string(),
});
export type FLEinladungEmpfaenger = z.infer<typeof FLEinladungEmpfaengerSchema>;

/**
 * Mirrors `FLEinladungVersandVorschauZeile` — one admitted team as the preview answers for it.
 * `empfaenger` is empty exactly where `uebersprungen` is set, so a row states either who would be
 * written to or why nobody would.
 */
export const FLEinladungVersandVorschauZeileSchema = z.object({
  team_id: CustomObjectIdStringSchema,
  // The season's own copy of the club's name, which is what the list is read and ordered by.
  team_name: z.string(),
  empfaenger: z.array(FLEinladungEmpfaengerSchema),
  uebersprungen: FLEinladungVersandGrundSchema.nullable(),
  // **The row carries no other trace of the invitation the team holds** — no id, no link value — so
  // nothing else here can tell a team whose live link this press kills from one that never had one.
  ersetzt_link: z.boolean(),
});
export type FLEinladungVersandVorschauZeile = z.infer<typeof FLEinladungVersandVorschauZeileSchema>;

/**
 * Mirrors `FLEinladungVersandZeile` — the preview's row as the press leaves it, carrying what the
 * message needs: the link's raw value, and the row a delivery report is filed against. Both are null
 * exactly where `uebersprungen` is set.
 */
export const FLEinladungVersandZeileSchema = FLEinladungVersandVorschauZeileSchema.extend({
  einladung_id: CustomObjectIdStringSchema.nullable(),
  token: z.string().nullable(),
  // Whether the team held a live link when the press read it; null where the press failed before reading.
  hatte_link: z.boolean().nullable(),
});
export type FLEinladungVersandZeile = z.infer<typeof FLEinladungVersandZeileSchema>;

/** Every ADMITTED team of the season, skipped ones included: a list that dropped them would report a season smaller than it is. */
export const FLEinladungVersandVorschauResponseSchema = BaseAPIResponseSchema.extend({
  saison_id: z.string(),
  zeilen: z.array(FLEinladungVersandVorschauZeileSchema),
});
export type FLEinladungVersandVorschauResponse = z.infer<typeof FLEinladungVersandVorschauResponseSchema>;

export const FLEinladungVersandResponseSchema = BaseAPIResponseSchema.extend({
  saison_id: z.string(),
  zeilen: z.array(FLEinladungVersandZeileSchema),
});
export type FLEinladungVersandResponse = z.infer<typeof FLEinladungVersandResponseSchema>;
